modules/xml/xml.c

```diff
@@ -0,0 +1,338 @@
+    {
+      msg_debug("xml: subtree skipped", evt_tag_str("tag", element_name));
+      state->pop_next_time = 1;
+      g_markup_parse_context_push(context, &skip, NULL);
+      return;
+    }
+
+  g_string_append_printf(state->key, ".%s", element_name);
+  msg_add_attributes(state->msg, state->key, attribute_names, attribute_values);
+}
+
+static gint
+scan_for_last_dot(GString *str)
+{
+  gint i = str->len;
+  while (str->str[i] != '.')
+    i--;
+  return i;
+}
+
+static void
+end_element_cb(GMarkupParseContext *context,
+               const gchar         *element_name,
+               gpointer             user_data,
+               GError              **error)
+{
+  InserterState *state = (InserterState *)user_data;
+
+  if (state->pop_next_time)
+    {
+      g_markup_parse_context_pop(context);
+      state->pop_next_time = 0;
+      return;
+    }
+  g_string_truncate(state->key, scan_for_last_dot(state->key));
+}
+
+static void
+text_cb(GMarkupParseContext *context,
+        const gchar         *text,
+        gsize                text_len,
+        gpointer             user_data,
+        GError             **error)
+{
+  if (text_len == 0)
+    return;
+
+  InserterState *state = (InserterState *)user_data;
+  const gchar *current_value = log_msg_get_value_by_name(state->msg, state->key->str, NULL);
+  GString *value = scratch_buffers_alloc();
+  g_string_printf(value, "%s%s", current_value, text);
+  log_msg_set_value_by_name(state->msg, state->key->str, value->str, value->len);
+}
+
+
+static GMarkupParser xml_scanner =
+{
+  .start_element = start_element_cb,
+  .end_element = end_element_cb,
+  .text = text_cb
+};
+
+static gboolean
+xml_parser_process(LogParser *s, LogMessage **pmsg,
+                   const LogPathOptions *path_options,
+                   const gchar *input, gsize input_len)
+{
+  XMLParser *self = (XMLParser *) s;
+  LogMessage *msg = log_msg_make_writable(pmsg, path_options);
+
+  GString *key = scratch_buffers_alloc();
+  key = g_string_append(key, self->prefix);
+
+  InserterState state = { .msg = msg, .key = key, .parser = self };
+  GMarkupParseContext *xml_ctx = g_markup_parse_context_new(&xml_scanner, 0, &state, NULL);
+
+  GError *error = NULL;
+  g_markup_parse_context_parse(xml_ctx, input, input_len, &error);
+  if (error)
+    goto err;
+
+  g_markup_parse_context_end_parse(xml_ctx, &error);
+  if (error)
+    goto err;
+
+  g_markup_parse_context_free(xml_ctx);
+  return TRUE;
+
+err:
+  msg_error("xml: error", evt_tag_str("str", error->message));
+  g_error_free(error);
+  g_markup_parse_context_free(xml_ctx);
+  return self->forward_invalid;
+}
+
+void
+xml_parser_set_exclude_tags(LogParser *s, GList *exclude_tags)
+{
+  XMLParser *self = (XMLParser *) s;
+
+  g_list_free_full(self->exclude_tags, g_free);
+  self->exclude_tags = g_list_copy(exclude_tags);
+
+  /* g_list_copy_deep would be better, but not supported on all glib versions :( */
```

Comment:
Would it be possible to implement g_list_copy_deep under a compiler switch (do we have g_list_copy_deep or not) ?
In this case the code would document this comment :)

Reply:
Thanks for the comment! I backported a g_list_copy_deep implementation to use on machines where it is not available. It is less efficient than the one provided by glib master, because this version iterates twice through the list instead of one. But it does not use any internal api from the glib list implementation (like memory allocation), so I felt safer to go with this version.